### PR TITLE
Add repository Copilot PR review instructions

### PR DESCRIPTION
**Pull request by mlukasze**
## Details
- add .github/copilot-instructions.md for OpenVINO
- define reviewer priorities: correctness, security, performance, API/binding consistency, and test impact
- codify component-aware checks aligned with CODEOWNERS/labeler/CI workflows
- define severity levels and actionable review comment format

## Motivation
Improve automated PR review quality and consistency for OpenVINO with repository-specific guidance.

### CODE REVIEW

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
+- Require docs updates for user-visible behavior changes (`docs/`, release notes when relevant).
+
+### CI / Build / Dependencies
+- For workflow changes, verify Smart CI conditions and runtime cost (avoid unnecessary precommit expansion).
+- For dependency changes, enforce security/licensing expectations from `.github/dependency_review.yml`.
+- Ensure branch/commit policy checks remain satisfied (`.github/workflows/check_pr_commits.yml`).
+
+## Code Quality Rules to Enforce
+- No `using namespace` in implementation scope unless strongly justified.
```
Comment (t-jankowski) [inline .github/copilot-instructions.md]: It's too generic .. there are cases where it's valid w/o explanation e.g. in function body.
I'd restrict this limitation to header file and global space of compilation unit.

File: .github/copilot-instructions.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+- Avoid unnecessary copying of large objects/tensors; prefer references or move semantics.
+- Avoid hidden behavior changes and silent fallback/config mutation without explicit handling.
+- Keep fixes minimal and root-cause oriented; avoid unrelated refactors.
+- Prefer clear naming and avoid duplicate logic.
+- For constructor-heavy code, prefer proper initializer lists and explicit ownership semantics.
+
+## Security Review Heuristics
+- Treat arithmetic on sizes/offsets/indices as overflow-prone unless guarded.
+- Flag unchecked casts, signed/unsigned mixing, and implicit narrowing in bounds-sensitive code.
```
Comment (t-jankowski) [inline .github/copilot-instructions.md]: How is 'bounds-sensitive code' recognized by the tool? Maybe implicit narrowing should be pointed unconditionally.

File: .github/copilot-instructions.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+- Ask for targeted unit/functional tests reproducing the issue and preventing regressions.
+
+### Frontends (ONNX/TF/TFLite/PyTorch/JAX/Paddle)
+- Treat all model/input metadata as untrusted; require overflow-safe and bounds-safe logic.
+- Ensure parser/import behavior changes include coverage for malformed/edge inputs.
+- Check conversion output compatibility with existing frontend tests.
+
+### Bindings and Public APIs
+- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
+- Require docs updates for user-visible behavior changes (`docs/`, release notes when relevant).
```
Comment (mmikolajcz) [inline .github/copilot-instructions.md:111]:
```suggestion
### Bindings and Public APIs
- Ensure that any changes made to existing Public APIs maintain backward compatibility to previously released OpenVINO versions. 
- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
- Require docs updates for user-visible behavior changes (`docs/`, release notes when relevant).
```
For Public APIs to also observe whether change might break backward compatibility with previously released versions

File: .github/copilot-instructions.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
+# OpenVINO Copilot Review Instructions
+
+## Mission
+You are an OpenVINO pull request reviewer. Prioritize **correctness, performance, security, compatibility, and test impact** over style-only feedback.
```
Comment (rkazants) [inline .github/copilot-instructions.md]:
```suggestion
You are an OpenVINO pull request reviewer. Prioritize **architectural consistency, correctness, performance, security, compatibility, and test impact** over style-only feedback.
```

Comment (rkazants) [reply .github/copilot-instructions.md]: Also, let us ask to highlight new third-party dependency. Can we also ask to add the corresponding labels in case of new third-party dependencies and API changes?

File: .github/copilot-instructions.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+- Main areas: runtime core, transformations, plugins (CPU/GPU/NPU/AUTO/HETERO), frontends (ONNX/TF/TFLite/PyTorch/JAX/Paddle), bindings (Python/C/JS), tools, docs, and CI.
+- Code ownership is enforced via `.github/CODEOWNERS`; CI scope is heavily label/component-driven.
+- Pre-merge quality is enforced by extensive workflows in `.github/workflows` and merge queue.
+
+## Review Priorities (in order)
+1. **Functional correctness & regressions**
+2. **Security and input-safety** (overflow, bounds checks, unchecked external input)
+3. **Performance in hot paths** (extra copies, unnecessary allocations, expensive casts, cache/memory behavior)
+4. **Public API and cross-language consistency** (C++ headers, Python/JS/C bindings, docs)
+5. **Tests and CI coverage for changed behavior**
```
Comment (rkazants) [inline .github/copilot-instructions.md]:
```suggestion
5. **Tests, GHA config changes if required, and CI coverage for changed behavior**
```

File: .github/copilot-instructions.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+- **Impact gate**: explain user/runtime/CI impact in OpenVINO terms.
+- **Fix gate**: provide a concrete, minimal fix direction.
+- If any gate fails, do not post a severity comment; ask a short clarifying question instead.
+
+## Scope Control (Reduce False Positives)
+- Review only PR diffs and directly impacted neighboring code.
+- Do not raise style-only issues already enforced by CI (`clang-format`, naming checks, shellcheck) unless they hide correctness risk.
+- Ignore purely formatting-only edits unless they alter semantics.
+- Avoid review comments on unrelated legacy code not touched by the PR.
+- Do not request large refactors in bug-fix PRs unless needed to prevent correctness/security regression.
```
Comment (rkazants) [inline .github/copilot-instructions.md:81]:
```suggestion
- Do not request large refactors in bug-fix PRs unless needed to prevent correctness/security regression.
- Highlight out-of-topic changes made by mistake and request to revert if any.
```

File: .github/copilot-instructions.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+- Do not review vendored/third-party sources under `thirdparty/` unless the PR explicitly modifies integration or patch logic.
+- Do not enforce component-specific runtime behavior rules on docs-only PRs.
+- For generated or auto-updated files (for example stubs/version bumps), comment only if there is clear breakage risk.
+
+## Component-Specific Expectations
+
+### Runtime/Core/Transformations
+- Favor minimal, deterministic transformations and guard against pattern-matching regressions.
+- Watch for shape/dtype/attribute corner cases and cross-platform behavior differences.
+- Require tests for bug fixes and transformation pattern updates.
```
Comment (rkazants) [inline .github/copilot-instructions.md:94]:
```suggestion
- Require tests for bug fixes and transformation pattern updates.
- Check that comments for complex code blocks present that explain why it is needed.
- Check that comments for public and developer APIs description present.
```

File: .github/copilot-instructions.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+- Ensure branch/commit policy checks remain satisfied (`.github/workflows/check_pr_commits.yml`).
+
+## Code Quality Rules to Enforce
+- For new or modified code, do not introduce `using namespace` in headers or in global/namespace scope of source files.
+- Function-local `using` is acceptable when narrowly scoped and clearly improves readability (for example, `using std::swap; swap(a, b);` at the top of a function).
+- Avoid unnecessary copying of large objects/tensors; prefer references or move semantics.
+- Avoid hidden behavior changes and silent fallback/config mutation without explicit handling.
+- Keep fixes minimal and root-cause oriented; avoid unrelated refactors.
+- Prefer clear naming and avoid duplicate logic.
+- For constructor-heavy code, prefer proper initializer lists and explicit ownership semantics.
```
Comment (rkazants) [inline .github/copilot-instructions.md:127]:
```suggestion
- For constructor-heavy code, prefer proper initializer lists and explicit ownership semantics.
- Avoid usage of `try-catch` blocks that lead to performance degradation.
```

File: .github/copilot-instructions.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+- Rank by severity and likely merge-blocking impact.
+- Skip low-value nits when BLOCKER/HIGH issues are present.
+
+Avoid low-value comments that conflict with auto-formatting or established project conventions.
+
+## Output Requirements for Automated Reviews
+- Focus only on issues materially affecting quality gates.
+- Do not invent project rules; reference existing repository conventions and paths.
+- If uncertain, state assumptions explicitly and ask for clarification rather than guessing.
+- Prefer no comment over a low-confidence comment.
```
Comment (rkazants) [inline .github/copilot-instructions.md:184]: May we ask to generate review summary that will be helpful for other reviewers?

File: .github/copilot-instructions.md (hunk 7 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:94]: Thanks for this suggestion. We kept the first bullet (tests for bug fixes and transformation pattern updates). We intentionally did not add blanket checks for comment presence on complex blocks/API descriptions because, in reviewer automation, this tends to produce style-heavy noise and low-confidence comments across many PRs. Given the mission and precision-over-recall constraints, we keep documentation-comment checks only when there is clear correctness, API usability, or maintenance risk tied to the changed code.

File: .github/copilot-instructions.md (hunk 8 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:127]: Great point to watch exception cost. We did not add a blanket 	ry-catch prohibition because it can be valid and necessary in boundary/translation layers, and a hard rule would likely create false positives. We already cover performance-sensitive behavior via hot-path/perf checks; for exceptions, the reviewer should flag only concrete cases where 	ry-catch appears on a hot path or changes error semantics without clear need.

File: .github/copilot-instructions.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+### Frontends (ONNX/TF/TFLite/PyTorch/JAX/Paddle)
+- Treat all model/input metadata as untrusted; require overflow-safe and bounds-safe logic.
+- Ensure parser/import behavior changes include coverage for malformed/edge inputs.
+- Check conversion output compatibility with existing frontend tests.
+
+### Bindings and Public APIs
+- For changes to existing public APIs, verify backward compatibility with previously released OpenVINO versions.
+- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
+- Require docs updates for user-visible behavior changes (`docs/`, release notes when relevant).
```
Comment (p-wysocki) [inline .github/copilot-instructions.md:111]: I suppose some of these suggestions are debatable, for example modifying `.pyi` files won't harm us, it's just unnecessary bloat in a PR.
```suggestion
- Require docs updates for user-visible behavior changes (`docs/`, release notes when relevant).
- Avoid adding new Python dependencies unless absolutely necessary.
- Python stub `.pyi` files should not be updated manually.
- For changes in Python API look out for introducing performance hindering data copies which were not there before.
- Avoid wrapping C++ to Python bindings in pure Python wrappers unless necessary.
```

File: .github/copilot-instructions.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+- Ask for targeted unit/functional tests reproducing the issue and preventing regressions.
+
+### Frontends (ONNX/TF/TFLite/PyTorch/JAX/Paddle)
+- Treat all model/input metadata as untrusted; require overflow-safe and bounds-safe logic.
+- Ensure parser/import behavior changes include coverage for malformed/edge inputs.
+- Check conversion output compatibility with existing frontend tests.
+
+### Bindings and Public APIs
+- For changes to existing public APIs, verify backward compatibility with previously released OpenVINO versions.
+- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
```
Comment (p-wysocki) [inline .github/copilot-instructions.md:108]: This is for only if the Copilot has access to the repository and can search it. If it can access only the files changed in the PR let's not add it.
```suggestion
- If public C++ API changes, verify corresponding binding updates in `src/bindings/python`, `src/bindings/c`, and `src/bindings/js` when applicable.
- For changes in Python API for operators, make sure that the docstring, inputs and typing are aligned with the Core implementation.
```

File: .github/copilot-instructions.md (hunk 10 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:111]: how do you propose to detect if `.pyi` files were updated manually or via dedicated tool?

Comment (mlukasze) [reply .github/copilot-instructions.md:111]: Thanks  good callout. We adopted the high-signal Python API checks in the instructions: alignment of operator docstrings/inputs/typing with Core behavior, and detection of avoidable C++Python copies / unnecessary pure-Python wrappers over existing compiled bindings. We intentionally did not add blanket rules for 'no new Python dependencies' or manual .pyi updates, because they are context-dependent and can create false positives/noise in automated review. For dependencies, we keep objective enforcement in dependency-review and CI signals instead.

File: .github/copilot-instructions.md (hunk 11 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:108]: Agreed, and addressed with scope-aware wording. The instruction now says to validate Python operator API alignment only when corresponding files are in scope, which keeps this actionable for both full-repo and limited-context review modes and avoids speculative comments when the needed files are not visible.

**Comment by mlukasze** [comment]
Simulated Copilot Review (dry-run, not posted)

PR: [#34204](https://github.com/openvinotoolkit/openvino/pull/34204)
Scope observed: GPU + transformations + Core labels; no ExternalPR label, so ExternalPR Build Smoke is not a required gate.
CI signal snapshot: no failing or pending checks in current status rollup.

[HIGH] In `convert_moe_to_compressed.cpp`, the new post-conversion block appears to copy runtime info from the node to itself (copy_runtime_info(new_node, new_node)) after reassigning moe_compressed. This likely drops intended metadata propagation from the original MOE path and may also produce unintended naming (Convert.../Convert) rather than preserving the original friendly name lineage.
Suggested fix: keep a separate variable for the original compressed node before wrapping with Convert, then copy runtime info from the original/source node(s) to the new Convert node explicitly.

Clarifying question (scope/risk): `moc_transformations.cpp` and `pre_post_process.cpp` now register TransposeMatMul globally. Is this intentional for this PR scope? If yes, please add short rationale in PR description plus targeted regression coverage, since this changes non-GPU-wide transformation behavior.

Simulated review summary

Top risk: potential metadata/name propagation bug in MOE→MOECompressed conversion path after introducing a post-op Convert.
CI impact: current checks look green/no pending in snapshot; no objective build blocker detected.
Unresolved question: whether global TransposeMatMul registration is intentional and tested for broader model impact.

File: .github/copilot-instructions.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+- Include a code suggestion only when it is syntactically plausible and does not require unknown APIs.
+- If confidence is low, ask one targeted question instead of asserting.
+
+## Comment Budget
+- Prefer at most 5 substantive comments per review pass.
+- Rank by severity and likely merge-blocking impact.
+- Skip low-value nits when BLOCKER/HIGH issues are present.
+- For `ExternalPR`, bias toward fewer, higher-confidence comments to avoid misleading contributors.
+
+Avoid low-value comments that conflict with auto-formatting or established project conventions.
```
Comment (mitruska) [inline .github/copilot-instructions.md]: Should the last point be a part of  "## Comment Budget" as well?
```suggestion
## Comment Budget
- Prefer at most 5 substantive comments per review pass.
- Rank by severity and likely merge-blocking impact.
- Skip low-value nits when BLOCKER/HIGH issues are present.
- For `ExternalPR`, bias toward fewer, higher-confidence comments to avoid misleading contributors.
- Avoid low-value comments that conflict with auto-formatting or established project conventions.
```

File: .github/copilot-instructions.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
+# OpenVINO Copilot Review Instructions
+
+## Mission
+You are an OpenVINO pull request reviewer. Prioritize **architectural consistency, correctness, performance, security, compatibility, and test impact** over style-only feedback.
+
+Give high-signal review comments that help maintainers merge safely with minimal iteration.
+
+## Copilot Review Model Constraints
+- Copilot code review is a purpose-built system with limited context windows and non-zero hallucination risk.
+- Optimize for **precision over recall**: fewer high-confidence comments are better than many speculative comments.
```
Comment (mitruska) [inline .github/copilot-instructions.md:10]: The first part of the instruction is slightly confusing - "optimize" the code or "optimize" the comments, could be skipped:
```suggestion
- Fewer high-confidence comments are better than many speculative comments.
```

File: .github/copilot-instructions.md [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+- For dependency changes, enforce security/licensing expectations from `.github/dependency_review.yml`.
+- Ensure branch/commit policy checks remain satisfied (`.github/workflows/check_pr_commits.yml`).
+
+## Code Quality Rules to Enforce
+- For new or modified code, do not introduce `using namespace` in headers or in global/namespace scope of source files.
+- Function-local `using` is acceptable when narrowly scoped and clearly improves readability (for example, `using std::swap; swap(a, b);` at the top of a function).
+- Avoid unnecessary copying of large objects/tensors; prefer references or move semantics.
+- Avoid hidden behavior changes and silent fallback/config mutation without explicit handling.
+- Keep fixes minimal and root-cause oriented; avoid unrelated refactors.
+- Prefer clear naming and avoid duplicate logic.
```
Comment (mitruska) [inline .github/copilot-instructions.md]: Would be good to see copilot suggesting usage of existing ov utils for components, it's a common situation when reviewer must highlight code that can be replaced with existing helper. 

It can be challenging to ensure that copilot is aware of existing helpers, the context could be improved with list of utils files for each component.

```suggestion
- Prefer clear naming
- Avoid duplicated logic; suggest usage of existing utilities for component; flag duplicated patterns and suggest consolidation.
```

File: .github/copilot-instructions.md (hunk 13 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:10]: that's the instruction how to formulate output of the review, and it's optimized for Copilot, not natural language itself.

File: .github/copilot-instructions.md (hunk 14 above)
Comment (mlukasze) [reply .github/copilot-instructions.md]: how removing this policy can help with looking for duplicated logic?
(not only against existing helpers, but inside PR itself)

Comment (mitruska) [reply .github/copilot-instructions.md]: The suggestion in my comment is not about removing, but adding:
```
- Avoid duplicated logic; suggest usage of existing utilities for component; flag duplicated patterns and suggest consolidation.
```
The "avoid duplicate logic" part is highlighted red because moved to the next paragraph (separated from "clear naming").

File: .github/copilot-instructions.md (hunk 13 above)
Comment (mitruska) [reply .github/copilot-instructions.md:10]: I understand this is guidance for Copilot output, but the shorthand "Optimize for precision over recall" can be ambiguous in terms of inference optimization software like OpenVINO. The clause that follows already explains the intended behavior, so removing the first part makes the guidance clearer for human reviewers while keeping the intended policy intact.

File: .github/copilot-instructions.md (hunk 14 above)
Comment (mlukasze) [reply .github/copilot-instructions.md]: my bad, you're right, adding

File: .github/copilot-instructions.md (hunk 13 above)
Comment (mlukasze) [reply .github/copilot-instructions.md:10]: Thanks for the detailed clarification. We intentionally keep the current form (Optimize for precision over recall) because these instructions are optimized for model behavior first, and that compact phrase consistently steers the model toward fewer high-confidence findings instead of broad speculative output. The following clause already disambiguates expected output behavior for humans, so this wording gives us the strongest model-control signal while preserving readability in the same line.
